Start applying ruff rules

DimitriPapadopoulos opened the pull request:
Start applying default ruff rules (`E` and `F`).

Fixes #447.

kevina commented:
Also, if possible please verify none of these rules will break Python 3.7 compatibility, otherwise I will have to check for myself and might require you to removal of a rule of two if it uses a newer Python 3 feature.

kevina commented:
> All the suggested changes are compatible with Python 3.7.

Thank you.  The next major upgrade will likely be Python 12, and Sqlite 3.44 (or maybe 3.45).  Python 12 for the improved transaction handling in sqlite3 and most importantly not auto commuting before running a script with `executescript`.  SQLite 3.44 for the improved aggregate functions.  However, that will create too much of a headache for those not using the latest distros.

So for now I am testing with Python 3.7 (what on my system) and Python 3.12.

DimitriPapadopoulos commented:
Most commits are automatic fixes, but I had to apply manual fixes in a couple cases to improve the code.

These special cases include of course [F841](https://docs.astral.sh/ruff/rules/unused-variable/#unused-variable-f841) / f9a9e9aa1d580dad33b71b73504f06e4d795362e, as I chose to remove actual (unused) code. Note that [unused-variable (F841)](https://docs.astral.sh/ruff/rules/unused-variable/#unused-variable-f841) is marked as "unsafe", so [`pre-commit.ci`](https://pre-commit.ci/) would probably not suggest automatic fixes in the commit appended to the PR:
> This rule's fix is marked as unsafe because removing an unused variable assignment may delete comments that are attached to the assignment.

With option `--fix --unsafe-fixes`, Ruff only suggests to prepend `_`.

In any case, fixes are not applied before you review and approve them: [`pre-commit.ci`](https://pre-commit.ci/) just appends it own commit to every PR, so your review would include the Ruff changes.

kevina commented:
@DimitriPapadopoulos I'm ready for this now.

Could you please re-run the rules? As before, please keep them as one commit per rule.

Once updated, I'll cherry-pick the rules I want to keep and squash them.  I will credit you via Co-authored-by regarding the email policy, just like I did for #464.

DimitriPapadopoulos committed:
Apply ruff/pycodestyle rule E703

Statement ends with an unnecessary semicolon

DimitriPapadopoulos committed:
Apply ruff/pycodestyle rule E714

Test for object identity should be `is not`

DimitriPapadopoulos committed:
Manually apply ruff/pycodestyle rule E722

Do not use bare `except`

DimitriPapadopoulos committed:
Apply ruff/Pyflakes rule F401

Module imported but unused

DimitriPapadopoulos committed:
Apply ruff/Pyflakes rule F541

f-string without any placeholders

DimitriPapadopoulos committed:
Manually apply ruff/Pyflakes rule F841

Local variable is assigned to but never used

DimitriPapadopoulos committed:
Apply ruff/flake8-comprehensions rule C401

Unnecessary generator (rewrite as a set comprehension)

DimitriPapadopoulos committed:
Apply ruff/flake8-comprehensions rule C416, reformat manually

Unnecessary dict comprehension (rewrite using `dict()`)

DimitriPapadopoulos committed:
Apply ruff/flake8-comprehensions rule C417, then UP034

Unnecessary `map()` usage (rewrite using a generator expression)
Avoid extraneous parentheses

DimitriPapadopoulos committed:
Apply ruff/flake8-simplify rules SIM114 and SIM101

Combine `if` branches using logical `or` operator
Multiple `isinstance` calls for `other`, merge into a single call

DimitriPapadopoulos committed:
Apply ruff/flake8-simplify rule SIM210

Remove unnecessary `True if ... else False`
Use `bool(...)` instead of `True if ... else False`

DimitriPapadopoulos committed:
Apply ruff/Perflint rule PERF403, then C416 again

Use a dictionary comprehension instead of a for-loop
Unnecessary dict comprehension (rewrite using `dict()`)

DimitriPapadopoulos committed:
Apply ruff/pyupgrade rules UP031 and UP032

Use format specifiers instead of percent format
Use f-string instead of `format` call

DimitriPapadopoulos committed:
Apply ruff/refurb rule FURB142

Use of `set.add()` in a for loop

DimitriPapadopoulos committed:
Apply ruff/flake8-pie rule PIE808

Unnecessary `start` argument in `range`

DimitriPapadopoulos committed:
Apply ruff/Perflint rule PERF401

Use `list.extend` to create a transformed list

DimitriPapadopoulos committed:
Apply ruff/refurb rule FURB110

Replace ternary `if` expression with `or` operator

DimitriPapadopoulos committed:
Apply ruff/refurb rule FURB113

Use `res.extend()` instead of repeatedly calling `res.append()`

DimitriPapadopoulos committed:
Apply ruff rule RUF027

Possible f-string without an `f` prefix

DimitriPapadopoulos committed:
Apply ruff rule RUF047

Empty `else` clause

kevina commented:
It's a very low priority right now.

I might have another look at both F481, F401 after I get a release out.  Feel free to open an issue or create another P.R. if you do create a P.R. please sanity test it with `make clean && make scowl.txt`.
